Small changes

hgloeckner opened the pull request:
- makes the featureType and corresponding cf_role in sonde_id more flexible. Closes #188 
- removes the alt_near_gpsalt test from the dataset. Closes #194 
- removes print if no attributes are in the config 
- changes the lat/lon to km numbers. Closes #159

tmieslinger reviewed (commented):
Thanks for adding those many small changes!
* I ran it on the example data and the `circle_id` ends up in the L4 dataset 👍
* I have some questions on the global attributes
* and I was wondering whether there is still a need to keep any code for `alt_near_gpsalt` in pydropsonde as I understood the changes such that they only make it not affect the dataset, but leave some code in pydropsonde.

hgloeckner commented on pydropsonde/processor.py:
```diff
@@ -738,11 +738,16 @@ def add_sonde_id_variable(self, variable_name="sonde_id"):
         }
+        if self.global_attrs["l2"].get("featureType") == "trajectoryProfile":
+            attrs.update(cf_role="trajectory_id")
+        elif self.global_attrs["l2"].get("featureType") == "profile":
+            attrs.update(cf_role="profile_id")
```
Not really, I just thought it would be nice to explicitly have this option there to in case the default is changed at some point. Maybe it's bad because it clutters the code?

tmieslinger replied:
hm. good point 👍

hgloeckner commented on pydropsonde/processor.py:
```diff
@@ -1575,12 +1580,20 @@ def add_ids(self):
         ds = self.interim_l3_ds
         source_ds = self.l2_ds
+
+        sonde_attrs = source_ds.sonde_id.attrs
+
+        if self.global_attrs["l3"].get("featureType") == "trajectoryProfile":
+            sonde_attrs.update(cf_role="trajectory_id")
+        elif self.global_attrs["l3"].get("featureType") == "profile":
+            sonde_attrs.update(cf_role="profile_id")
+
```
I would say that people should specify this in the config. At the moment the feature type is never inherited. The default is "profile" for level 2 and "trajectoryProfile" for level 3 and 4. I would suggest that people have to specify in the config if they want anything other than that (preferably specify it always)

hgloeckner commented:
- added the cosmetic changes to level 4 as mentioned in Issue #200

tmieslinger reviewed (commented):
Thanks for adding the small changes from issue #200! I went once through the code and made some comments inline. But here is a list of things that I noted while going through all variables in the processed L4 dataset. Maybe not all are super relevant:
* The title still has the underscore “Level_4” and could be replaced by a space
* `history` global attribute: does “level3 concatenation” need to be listed separately? Without knowing the details, it might be confusing to read “Level 3 processing” followed by “Level 3 concatenation”. Independent, we could make it easier to read by adding “Level 1” to the ASPEN processing item, “Level 2” to the QC item and leave out the date time for Level 3 concatenation and Level 4 computation?
* `altitude` still has the non-existing ancillary_variables `altitude_below_aircraft` and `altitude_source`. Also, the description says “derived from either gpsalt or alt in Level 2. See alt_source” and could be adjusted to e.g. “best estimate from either GPS or pressure measurements. See details for each sonde in respective Level 2 dataset.” (Maybe a bit long)
* `circle_time` could get `time_zone: UTC` similar to `sonde_time`
* `sondes_per_cricle` should get a description attribute, e.g. “Number of sondes per circle. Count variable for dimension sonde.”
* `div_std_error` should have the same unit as `div`, right? Currently it has `m s-1 m-1` (instead of `s-1`?)
* A very small thing: `iwv` has the spelling “vapor” in the standard name, but “vapour” in description, lon_name and also in `iwv_mean` long_name. We could adjust the spelling.
* A few variables have standard_names that I am unsure whether they are correct or should rather be long_names: div_std_error, omega_std_error, u_dudx_std_error, u_dudy_std_error, v_dvdx_std_error, v_dvdy_std_error, wvel_std_error
* `q`and `theta`still have ancillary variables `?_N_qc` that are not in the ds anymore (see also inline comment)
* `sonde_id` gets the `cf_role: profile_id` in L4 similar to L3 (just like we discussed). Thinking about it, it would equally make sense to me to give `circle_id` the `cf_role: profile_id` as this would fit to all of the circle products (div, omega…) and that’s the purpose of L4. Any opinions here @helene and @lukas?
* `vor` gets the ancillary_variable `div_standard_error`. Is this intended?

hgloeckner commented on pydropsonde/circles.py:
```diff
@@ -340,72 +340,74 @@ def add_regression_stderr(self, variables=None):
-        ds = hx.add_ancillary_var(ds, "vor", "se_div")
-        se_div_nona = ds.se_div.dropna(dim=alt_dim)
+        ds = hx.add_ancillary_var(ds, "div", "div_standard_error")
+        ds = hx.add_ancillary_var(ds, "vor", "div_standard_error")
```
This was wrong - also the calculation for the divergence standard error and vorticity standard error were wrong. thanks for pointing this out!

hgloeckner commented on pydropsonde/circles.py:
```diff
@@ -340,72 +340,74 @@ def add_regression_stderr(self, variables=None):
                     [alt_dim],
-                    np.sqrt(ds.se_dudx**2 + ds.se_dudy**2).values,
+                    np.sqrt(ds.u_dudx_std_error**2 + ds.u_dudy_std_error**2).values,
                     dict(standard_name=f"{div_std_name} standard_error", units=unit),
```
if I interpret the [cf convention](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.12/cf-conventions.html#standard-name) 3.4 correctly, then yes

tmieslinger replied:
True! I had in mind the general definition of a standard_name which shall not include whitespace. But seems like in this case of ancillary variables your implementation is exactly according to cf conventions. Thanks for teaching me new stuff :)

hgloeckner commented on pydropsonde/circles.py:
```diff
@@ -50,29 +55,38 @@ def drop_vars(self, variables=None):
             )
             .drop_vars(
-                [f"{var}_qc" for var in ["u", "v", "ta", "p", "rh"]],
+                [f"{var}_qc" for var in data_vars],
```
thanks for pointing this out!

hgloeckner commented:
Thank you for the comments; 
> * The title still has the underscore “Level_4” and could be replaced by a space

changed to be space by default (however, the recommended way would be to define the titles in the config)

> * `history` global attribute: does “level3 concatenation” need to be listed separately? Without knowing the details, it might be confusing to read “Level 3 processing” followed by “Level 3 concatenation”. Independent, we could make it easier to read by adding “Level 1” to the ASPEN processing item, “Level 2” to the QC item and leave out the date time for Level 3 concatenation and Level 4 computation?

because of the interim_l3 structure it is possible to have the calculation and concatenation from different versions of pydropsonde (right now it's not, but theoretically it could be possible). That's why there are two entries. 
Also, the dates are there because of cf-conventions. but we can also delete them from Level 3 and 4 (for the earlier steps they have to be deleted anyways)

> * `altitude` still has the non-existing ancillary_variables `altitude_below_aircraft` and `altitude_source`. Also, the description says “derived from either gpsalt or alt in Level 2. See alt_source” and could be adjusted to e.g. “best estimate from either GPS or pressure measurements. See details for each sonde in respective Level 2 dataset.” (Maybe a bit long)

Thanks! that's wrong and changed

> * `circle_time` could get `time_zone: UTC` similar to `sonde_time`

added!

> * `sondes_per_cricle` should get a description attribute, e.g. “Number of sondes per circle. Count variable for dimension sonde.”

done!

> * `div_std_error` should have the same unit as `div`, right? Currently it has `m s-1 m-1` (instead of `s-1`?)

that's right

> * A very small thing: `iwv` has the spelling “vapor” in the standard name, but “vapour” in description, lon_name and also in `iwv_mean` long_name. We could adjust the spelling.

Done

> * A few variables have standard_names that I am unsure whether they are correct or should rather be long_names: div_std_error, omega_std_error, u_dudx_std_error, u_dudy_std_error, v_dvdx_std_error, v_dvdy_std_error, wvel_std_error

see comment above

> * `q`and `theta`still have ancillary variables `?_N_qc` that are not in the ds anymore (see also inline comment)

see comment above.

> * `sonde_id` gets the `cf_role: profile_id` in L4 similar to L3 (just like we discussed). Thinking about it, it would equally make sense to me to give `circle_id` the `cf_role: profile_id` as this would fit to all of the circle products (div, omega…) and that’s the purpose of L4. Any opinions here @Helene and @Lukas?

I'd maybe leave it with the sonde_id. but I cannot really say why^^'

> * `vor` gets the ancillary_variable `div_standard_error`. Is this intended?

see above. there was an error in the std error calculation and assignment.

tmieslinger commented on pydropsonde/circles.py:
```diff
@@ -340,72 +340,74 @@ def add_regression_stderr(self, variables=None):
+                    .broadcast_like(ds[div_error_name])
                     .values,
                     dict(
                         standard_name=f"{wvel_std_name} standard_error",
```
not relevant. see conversation above.

tmieslinger commented on pydropsonde/processor.py:
```diff
@@ -2325,11 +2321,7 @@ def get_l4_filename(self, l4_filename: str = None):
-            + datetime.now(timezone.utc).isoformat()
-            + f" level4 computation with pydropsonde {__version__} \n"
-        )
+        history = history + f"Level4 computation with pydropsonde {__version__} \n"
```
maybe add a whitespace in `Level4` to be consistent with the other history items?

hgloeckner replied:
I think I fixed it

tmieslinger commented on pydropsonde/processor.py:
```diff
@@ -354,7 +354,7 @@ def add_aspen_history(self):
             history
             + aspen_time.isoformat()
-            + f" ASPEN processing with {aspen_version} \n"
+            + f"Level 1 ASPEN processing with {aspen_version} \n"
```
I don't know the reason for it, but the string needs to start with a whitespace here. Otherwise, the "Level" is swallowed and doesn't end up in the written file.
Same thing for the below changes in Level 2 and Level 3 processing. 
The history entries on Level 3 concatenation and Level 4 computation are fine.

hgloeckner replied:
it's because in earlier levels there is a date before that has to be removed and the string separator is a whitespace. maybe it's not the best solution...

hgloeckner replied:
but now it should work^^

tmieslinger reviewed (commented):
Thanks for adding everything so quickly.
I have one last thing that I think I forgot to mention earlier: `div` has the ancillary_variable `div_standard_error`, but the actual variable in the dataset is named `div_std_error`. Same for `vor`. I think you specify that [here?](https://github.com/atmdrops/pydropsonde/pull/195/commits/87063e98267bdf4163293e93d1f991f25c7defdb#diff-de6517483ebbcb7de63844b4774eab521a5e54197cd5d445047293139b1ebfc9R456-R457)
